### DISCUSSION TITLE
Replacement for MyV2xMessage()

### DISCUSSION THREAD

**Discussion by pv2-0** [eclipse/mosaic#58] [Q&A] [answered]
I am trying to have my vehicles broadcast their location at every timestep. I am basing that part of my application off of[ this example](https://www.eclipse.org/mosaic/docs/develop_applications/communication/) which suggests:
```
GeoPoint pointA = GeoPoint.latlon(52.51355, 13.22000);
GeoPoint pointB = GeoPoint.latlon(52.52000, 13.21000);
GeoRectangle transmissionArea = new GeoRectangle(pointA, pointB);

MessageRouting routing = getOs().getCellModule().createMessageRouting().geoBroadCast(transmissionArea);

getOs().getCellModule().sendV2XMessage(new MyV2XMessage(routing));
```

In order to get it to compile, I had to replace the geoBroadCast() function with geoBroadcastMbms(). After this change it will compile if I commend out the last line. Now I am having trouble with MyV2XMessage() which I cannot find any information on. I am wondering if the name of this class has changed? It seems to me like I should be using V2xMessage(), which is described [here](https://www.eclipse.org/mosaic/java_docs/org/eclipse/mosaic/lib/objects/v2x/V2xMessage.html#V2xMessage-org.eclipse.mosaic.lib.objects.v2x.MessageRouting-):

I tried using its subclass GenericV2xMessage() in the place of MyV2XMessage() in the code above, but I'm told I'm missing an argument "ToDataOutput". This argument isn't mentioned in the page describing GenericV2xMessage() at all:

> ...constructor GenericV2xMessage in class org.eclipse.mosaic.lib.objects.v2x.GenericV2xMessage cannot be applied to given types;
> [ERROR] required: org. eclipse.mosaic.lib.objects.v2x.MessageRouting,org.eclipse.mosaic.lib.objects.ToDataOutput,long
> [ERROR] found: org. eclipse.mosaic.lib.objects.v2x.MessageRouting
> [ERROR] reason: actual and formal argument lists differ in length

Does MyV2xMessage still exist? If not, what is the best way to fix the example above to broadcast the vehicle's location? If GenericV2xMessage() is the right replacement function, how do I use it?

**Comment by pv2-0** [comment, marked as answer]
The closest I got was doing:
`getOs().getCellModule().sendV2xMessage(new MyV2XMessage.Empty(routing));`
(the x in sendV2xMessage() should be lowercase!)
I still don't understand how to use ToDataOutput though.

**Reply by kschrab** [reply]
`ToDataOutput` is an interface which the class of the object you want to encode with needs to implement. For example, the `CamContent` class is the payload of the `Cam` message (extends `V2xMessage) and does implement this interface. 

Usually you do **not** need to encode the content of the message as for communication simulation only the length of the message is important. Instead of using `new V2xMessage.Empty(routing)` you could also create an own V2xMessage class with specifying the length of the message's content:

```java
public class MyV2xMessage extends V2xMessage {

        final int messageLength;

        public MyV2xMessage(MessageRouting routing, int messageLength) {
            super(routing);
            this.messageLength = messageLength;
        }

        @Nonnull
        @Override
        public EncodedPayload getPayLoad() {
            return new EncodedPayload(messageLength);
        }

}
```
